temporal-sdk/src/main/java/io/temporal/internal/client/RootWorkflowClientInvoker.java

```diff
@@ -356,20 +349,110 @@ public <R> UpdateOutput<R> update(UpdateInput<R> input) {
-      future.complete(update(input).getResult());
-    } catch (Exception e) {
-      future.completeExceptionally(e);
-    }
-    return new UpdateAsyncOutput<>(future);
+  public <R> PollWorkflowUpdateOutput<R> pollWorkflowUpdate(PollWorkflowUpdateInput<R> input) {
+    DataConverter dataConverterWithWorkflowContext =
+        clientOptions
+            .getDataConverter()
+            .withContext(
+                new WorkflowSerializationContext(
+                    clientOptions.getNamespace(), input.getWorkflowExecution().getWorkflowId()));
+
+    UpdateRef update =
+        UpdateRef.newBuilder()
+            .setWorkflowExecution(input.getWorkflowExecution())
+            .setUpdateId(input.getUpdateId())
+            .build();
+
+    WaitPolicy waitPolicy =
+        WaitPolicy.newBuilder()
+            .setLifecycleStage(
+                UpdateWorkflowExecutionLifecycleStage
+                    .UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED)
+            .build();
+
+    PollWorkflowExecutionUpdateRequest pollUpdateRequest =
+        PollWorkflowExecutionUpdateRequest.newBuilder()
+            .setNamespace(clientOptions.getNamespace())
+            .setIdentity(clientOptions.getIdentity())
+            .setUpdateRef(update)
+            .setWaitPolicy(waitPolicy)
+            .build();
+
+    CompletableFuture<PollWorkflowExecutionUpdateResponse> future = new CompletableFuture<>();
+
+    Deadline pollTimeoutDeadline = Deadline.after(input.getTimeout(), input.getTimeoutUnit());
+    pollWorkflowUpdateHelper(future, pollUpdateRequest, pollTimeoutDeadline);
+    return new PollWorkflowUpdateOutput(
+        future.thenApply(
+            (result) -> {
+              if (result.hasOutcome()) {
+                switch (result.getOutcome().getValueCase()) {
+                  case SUCCESS:
+                    Optional<Payloads> updateResult = Optional.of(result.getOutcome().getSuccess());
+                    R resultValue =
+                        convertResultPayloads(
+                            updateResult,
+                            input.getResultClass(),
+                            input.getResultType(),
+                            dataConverterWithWorkflowContext);
+                    return resultValue;
+                  case FAILURE:
+                    throw new WorkflowUpdateException(
+                        input.getWorkflowExecution(),
+                        input.getUpdateId(),
+                        input.getUpdateName(),
+                        dataConverterWithWorkflowContext.failureToException(
+                            result.getOutcome().getFailure()));
+                  default:
+                    throw new RuntimeException(
+                        "Received unexpected outcome from poll update request: "
+                            + result.getOutcome().getValueCase());
+                }
+              }
+              throw new RuntimeException("Received no outcome from server");
```

Comment:
If I'm understanding correctly, this RuntimeException would be if the future _didn't_ complete exceptionally but also _did_ return a null value for Outcome. So basically if the server broke contract, is that right?

Reply:
Yes exactly